[ISSUE-302] Sync website notes with youtube

azihassan opened the pull request:
Relevant issue : https://github.com/DevC-Casa/geeksblabla.com/issues/302

yjose commented on .github/workflows/upload_description.yaml:
```diff
@@ -0,0 +1,53 @@
+        run: |
+          echo "Changed files"
+          echo ${{ needs.detect_changes.outputs.all }}
+          for changed_file in ${{ needs.detect_changes.outputs.all }}; do
```
I think using a script shell to get episode youtube URL is a little bit risky here, I would suggest using javascript for this may be using the same approach we did with the podcast script. 

here is my suggestion: 
```yml

- name: Extract descriptions and URLs
        id: extract_descriptions_and_urls
        env:
          CLIENT_ID: ${{ secrets.CLIENT_ID }}
          CLIENT_SECRET: ${{ secrets.CLIENT_SECRET }}
          REDIRECT_URL: ${{ secrets.REDIRECT_URL }}
          ACCESS_TOKEN: ${{ secrets.ACCESS_TOKEN }}
          REFRESH_TOKEN: ${{ secrets.REFRESH_TOKEN }}
          UPDATED_FILES: ${{ needs.detect_changes.outputs.all }}
        run:  yarn upload-description-to-youtube 
```

In this case, you will have all updated files as env variables and you can play with them in js script, then you can create a simple script to get episode details using `front-matter` module  like the following :  
```js
// get-episode-details.js

const getEpisodeDetails = (file) => {
  try {
    const fileContent = fs.readFileSync(file, "utf8")
    const description = markdownToTxt(fileContent)
    // read file meta
    const meta = fm(fileContent)
    const youtubeUrl = meta.attributes.youtube // you can get the id directly here
    const episode = { youtubeUrl, description }
    return episode
  } catch (error) {
    return null
  }
}


```

azihassan replied:
Noice, TIL about front-matter.

I moved the shell logic to javascript in this commit : https://github.com/DevC-Casa/geeksblabla.com/pull/305/commits/7be875d203902593179d771da8cbe600d60d4526

But I didn't catch the exception inside it. Instead, I let it throw exceptions then handled them in the loop. Is that cool ?

yjose commented on youtube/get-tokens.js:
```diff
@@ -0,0 +1,24 @@
+const { google } = require("googleapis")
+
+const authorize = async ({
```
I Woudl like it if you can return the `auth` instead of a function that we need to call to get it. i think it would be more clear and clean with this way we will make sure we are using the same auth instance even if we need to update multiple videos

azihassan replied:
I see what you mean. I'm inclined to perform the authentication step in get-tokens.js and export the resulting oauth2Client, but how do we make sure it won't do this every time get-tokens is imported ?

This is what I had in mind :

```javascript
const { google } = require("googleapis")

const authorize = async () => {
  console.log("Logging in...")
  const oauth2Client = new google.auth.OAuth2(
    process.env.CLIENT_ID,
    process.env.CLIENT_SECRET,
    process.env.REDIRECT_URL
  )

  oauth2Client.setCredentials({
    access_token: process.env.ACCESS_TOKEN,
    refresh_token: process.env.REFRESH_TOKEN,
    scope: "https://www.googleapis.com/auth/youtube.force-ssl",
  })
  return oauth2Client
}
const auth = authorize()

module.exports = auth
```

Then in index.js :

```javascript
(async () => {
  const auth = await require("./get-tokens")
  const youtubeUrl = process.argv[2]
```

This however leaves a bad taste in my mouth. The module makes some assumptions that the consumers of the module might not know about, like defining env variables and having to await it during import.

yjose replied:
In this case, i would recommend creating a function `getAuth` that returns the auth instance and in case we already have an instance we return it ( singleton pattern )

so `getAuth` should look like the following:

```js
const { google } = require("googleapis")
let auth = null;

const getAuth = async () => {
    if(auth!==null){
     auth =  new google.auth.OAuth2(
    process.env.CLIENT_ID,
    process.env.CLIENT_SECRET,
    process.env.REDIRECT_URL
  )

  auth.setCredentials({
    access_token: process.env.ACCESS_TOKEN,
    refresh_token: process.env.REFRESH_TOKEN,
    scope: "https://www.googleapis.com/auth/youtube.force-ssl",
  })
}
return auth 
}
```

yjose commented on youtube/quickstart.js:
```diff
@@ -0,0 +1,131 @@
+// If modifying these scopes, delete your previously saved credentials
+// at ~/.credentials/youtube-nodejs-quickstart.json
+var SCOPES = ["https://www.googleapis.com/auth/youtube.force-ssl"]
+var TOKEN_DIR =
```
Can we update the path here to generate credentials inside youtube folder instead of root

azihassan replied:
Understood, just pushed an update here : https://github.com/DevC-Casa/geeksblabla.com/pull/305/commits/ef8ea248a8e9d5da91ff7b3536f0595c327920c7

I should mention that this file is a slightly modified copy of [this one](https://developers.google.com/youtube/v3/quickstart/nodejs), hence the callback hell and outdated practices

yjose reviewed (changes requested):
Great work so far @azihassan 🙏  i recommended some updates, let me know if you have any questions

yjose commented:
@azihassan   i did some refactoring on how we generate text description from markdown episode, feel free to review it and give me your inputs

azihassan commented:
Nice, I'm on it

yjose commented:
@azihassan  i am available starting from 10 pm to 11 pm, so we can do a call to finalize the workflow. 
Let me know if you are available

azihassan commented:
Hmm hard to say, 10 PM+ is difficult for me to guarantee. How about 9 to 10 PM ?

yjose commented:
@azihassan  Nicely done; I did some updates and generated youtube tokens. The guide looks very clean and easy to follow. 

Great work, i manage to update episode 98, and it looks excellent https://www.youtube.com/watch?v=39M0RTjPMnA 

Right now we should update all notes as we are not adding seconds to notes which seem to not work correctly with youtube

azihassan commented:
Glad to hear it ! Now that it works, I should remote the 1337 and 1338 episodes since I used them for testing. I'll do it as soon as I switch to my laptop.

azihassan commented:
Changed `steps.files.outputs.all` to `steps.files.outputs.added_modified` because I noticed that the action got triggered after I deleted two episode files.
